src/api/java/baritone/api/pathing/goals/GoalPatrol.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * This file is part of Baritone.
+ *
+ * Baritone is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Baritone is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Baritone.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package baritone.api.pathing.goals;
+
+import baritone.api.BaritoneAPI;
+import baritone.api.IBaritone;
+import baritone.api.process.ICustomGoalProcess;
+import baritone.api.utils.interfaces.IGoalRenderPos;
+import net.minecraft.util.math.BlockPos;
+
+import java.util.List;
+
+public class GoalPatrol implements Goal, IGoalRenderPos {
+
+    private List<Goal> waypoints;
+    private Mode mode;
+    private int index;
+
+    public GoalPatrol(List<Goal> waypoints, Mode mode, int index) {
+        this.waypoints = waypoints;
+        this.mode = mode;
+        this.index = index;
+    }
+
+    @Override
+    public boolean isInGoal(int x, int y, int z) {
+        boolean inGoal = waypoints.get(index).isInGoal(x, y, z);
+        BlockPos playerPos = BaritoneAPI.getProvider().getPrimaryBaritone().getPlayerContext().playerFeet();
+        if (playerPos.getX() == x && playerPos.getY() == y && playerPos.getZ() == z) {
+            if (inGoal) {
+                return getNextOrReturnTrue();
+            }
+            return false;
+        } else {
+            return inGoal;
+        }
+    }
```

Comment:
Sorry, to say it this harshly but NO.
1. I have concerns about thread safety (`playerFeet()` accesses the world)
2. This method is performance critical
3. `Goal`s are assumed to be stateless (and definitely should not depend on game state)

If you want to add patrolling you'll have to make some process responsible for changing the goal.

Reply:
I know (at least now, 27 minutes after submitting my previous review. Please forgive me) this pr is a draft, but since this is a pretty fundamental structural problem with this approach to patrolling I'll write it anyway.
If you want a changing goal you need some process to run the logic, both because goals are assumed to be stateless and because making them depend on game state like this can cause threading and performance problems.